realm/realm-library/src/main/java/io/realm/ObjectChangeSet.java

```diff
@@ -0,0 +1,49 @@
+/*
+ * Copyright 2017 Realm Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.realm;
+
+/**
+ * Information about the changes made to an object.
+ * @see RealmObject#addChangeListener(RealmObjectChangeListener) .
+ */
+public interface ObjectChangeSet {
+
+    /**
+     * Information about a specific field which changed in an {@link RealmObject} change notification.
+     */
+    class FieldChange {
+        /**
+         * The name of the field which changed.
+         */
+        final String fieldName;
+
+        public FieldChange(String fieldName) {
+            this.fieldName = fieldName;
+        }
+    }
+
+    /**
+     * @return {@code null} if the object has been deleted. Otherwise returns the information about changed fields.
+     */
+    FieldChange[] getFieldChanges();
```

Comment:
This wouldn't very well if you had e.g an `int` and a `String` field, as this would basically be type-erased to `FieldChange<Object>`.

Figuring out how to represent the changes are probably the crux of the problem, since the rest of the API's fall fairly natural from what we did for Collection notifications. Also, we probably need to consider what is most likely if interested in multiple changes. Do you register 1 listener and do switches or register multiple listeners?

What about this, it adds a few more helper methods people can use, that could provide a little more type safety. 

```
public interface ObjectChangeset {
  boolean isDeleted();
  // I'm not sure this is very useful in practical code. I cannot think of any generic code
  // where you wouldn't need the type information somehow?
  FieldChange[] getChanges(); 
  String[] getFieldNames;
  boolean didChange(String fieldName);
  <E> FieldChange<E> getChange(E type, String fieldName);

   /// Alternatively
   FieldChange<Long> getLongChange(String fieldName);
   FieldChange<String> getStringChange(String fieldName);
   FieldChange<Double> getDoubleChange(String fieldName);
   FieldChange<Float> getFloatChange(String fieldName);
   FieldChange<Boolean> getBooleanChange(String fieldName);
   FieldChange<Date> getDateChange(String fieldName);
   <E> FieldChange<E> getRealmObjectChange(E type, String fieldName);
   <E> FieldChange<RealmList<E>> getRealmListChange(E type, String fieldName);
}
```
Generally, I think we should try to make it as easy as possible for people to get the typed information since that is most likely what you want to operate in client code.